packages/contract/src/pay/DaimoPay.sol

```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.12;
+
+import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import "openzeppelin-contracts/contracts/utils/ReentrancyGuard.sol";
+
+import "../../vendor/cctp/ICCTPReceiver.sol";
+import "../../vendor/cctp/ICCTPTokenMessenger.sol";
+import "../../vendor/cctp/ITokenMinter.sol";
+import "./DaimoPayBridger.sol";
+import "./PayIntent.sol";
+import "./PayIntentFactory.sol";
+import "./TokenUtils.sol";
+
+// A Daimo Pay transfer has 4 steps:
+// 1. Alice sends (tokenIn, amountIn) to handoff address on chain A -- simple erc20 transfer
+// 2. LP swaps tokenIn to bridgeTokenIn and burns on chain A -- LP runs this in sendAndSelfDestruct
+//    - LP doesnt need approve or anything else, just quote and swap.
+//    - quote comes from the handoff address which commits to the destination bridgeTokenOut amount, and therefore bridgeTokenIn amount.
+//    - LP has to fetch swap from Uniswap or similar
+
+// Fork: fastFinish, then claim
+// Fork: claim directly
+
+// 4. LP swaps bridgeTokenOut to tokenOut on chain B -- LP runs this in completeAction
+// Alice is responsible for putting a quote for the bridgeTokenOut <> tokenOut swap
+// This fixes bridgeTokenOut expected amount, which in turn fixes the bridgeTokenIn burn amount,
+// locking in the amounts expected for all intermediary swaps.
+
+// To be nice Alice can put a a slightly worse quote than the market price to incentive LP.
+
+/// @title Daimo Pay contract for creating and fulfilling cross-chain actions
+/// @author The Daimo team
+/// @custom:security-contact security@daimo.com
+///
+/// Allows optimistic fast actions. Alice initiates a transfer by calling
```

Comment:
`@dev  ...`

Reply:
we should note up here that nobody should ever transfer or approve tokens to this contract as a standalone transaction. anyone can steal such tokens.

instead:
* users only interact by sending funds into an intent address
* relayers call this contract via their own contracts. for fastFinish..., the relayer calls via an atomic transaction that provides funds + calls fastFinish, without ever leaving a dangling transfer or approval.

this contract behaves analagously to the Uniswap router: you can atomically send funds to the router + swap, but if you should not send funds to the router in a standalone tx (say, from an EOA).